internal/server/deviceflow.go

```diff
@@ -64,46 +64,105 @@ retry:
 	if err != nil {
-		return nil, err
+		if errors.Is(err, internal.ErrNotFound) {
+			return nil, err
+		}
+		return nil, fmt.Errorf("%w: error retrieving device flow auth request: %v", internal.ErrUnauthorized, err)
 	}
 
 	if dfar.ExpiresAt.Before(time.Now()) {
 		return &api.DeviceFlowStatusResponse{
-			Status:     "expired",
+			Status:     api.DeviceFlowStatusExpired,
 			DeviceCode: dfar.DeviceCode,
 		}, nil
 	}
 
-	if dfar.AccessKey != nil {
+	if !dfar.Approved() {
 		return &api.DeviceFlowStatusResponse{
-			Status:     "confirmed",
+			Status:     api.DeviceFlowStatusPending,
 			DeviceCode: dfar.DeviceCode,
-			LoginResponse: &api.LoginResponse{
-				UserID:           dfar.AccessKey.IssuedFor,
-				Name:             dfar.AccessKey.IssuedForName,
-				AccessKey:        string(dfar.AccessKeyToken),
-				Expires:          api.Time(dfar.AccessKey.ExpiresAt),
-				OrganizationName: dfar.Organization.Name,
-			},
 		}, nil
 	}
 
+	user, err := data.GetIdentity(rctx.DBTxn, data.GetIdentityOptions{ByID: dfar.UserID})
+	if err != nil {
+		return nil, fmt.Errorf("%w: retrieving approval user: %v", internal.ErrUnauthorized, err)
+	}
+
+	// TODO : there's a lot of repeated logic in here and the Login() endpoint
+	// The Login() endpoint implemention needs to be re-worked into more atomic pieces that can
+	// be re-used here for setting cookies, creating an access key and a LoginResponse
+	accessKey := &models.AccessKey{
+		IssuedFor:     user.ID,
+		IssuedForName: user.Name,
+
+		// Share the same provider ID that was used to approve
+		ProviderID:          dfar.ProviderID,
+		ExpiresAt:           time.Now().UTC().Add(a.server.options.SessionDuration),
+		InactivityTimeout:   time.Now().UTC().Add(a.server.options.SessionInactivityTimeout),
+		InactivityExtension: a.server.options.SessionInactivityTimeout,
+		Scopes:              models.CommaSeparatedStrings{models.ScopeAllowCreateAccessKey},
+	}
+
+	bearer, err := data.CreateAccessKey(rctx.DBTxn, accessKey)
+	if err != nil {
+		return nil, fmt.Errorf("%w: creating new access key: %v", internal.ErrUnauthorized, err)
+	}
+
+	user.LastSeenAt = time.Now().UTC()
+	if err := data.UpdateIdentity(rctx.DBTxn, user); err != nil {
+		return nil, fmt.Errorf("%w: update user last seen: %v", internal.ErrUnauthorized, err)
+	}
+
+	cookie := cookieConfig{
+		Name:    cookieAuthorizationName,
+		Value:   bearer,
+		Domain:  c.Request.Host,
+		Expires: accessKey.ExpiresAt,
+	}
+	setCookie(c, cookie)
+
+	a.t.User(accessKey.IssuedFor.String(), user.Name)
+	a.t.OrgMembership(accessKey.OrganizationID.String(), accessKey.IssuedFor.String())
+	a.t.Event("login", accessKey.IssuedFor.String(), accessKey.OrganizationID.String(), Properties{"method": "deviceflow"})
+
+	// Update the request context so that logging middleware can include the userID
+	rctx.Authenticated.User = user
+	c.Set(access.RequestContextKey, rctx)
+
```

Comment:
These 2 lines will make it so that the log line emitted for this request includes a `userID`. Without this the userID is not set because the request is not authenticated until the device code is approved.